ТЗ-12

IIExpanse opened the pull request:
Андрей, добрый день. Высылаю ТЗ-12 на проверку.
В корневой директории лежит изображение "Filmorate ER-Diagram.png" - схема БД.
Я не проводил ни с кем совместных проверок и одна итерация уходит на проверку схемы, насколько я понял из ТЗ-11.

IIExpanse committed:
Добавлена БД. Логика работы приложения переделана под работу с БД. Старые классы и методы для работы в оперативной памяти сохранены, где возможно.

IIExpanse committed:
Попытка починить тесты 1

IIExpanse committed:
Попытка починить тесты 2

IIExpanse committed:
Попытка починить тесты 3

IIExpanse committed:
Попытка починить тесты 4

Nagash40 commented:
С тестами постмана в гите какая-то проблема, на все приходит 500 ошибка.
Локально все тесты проходят: и постмана, и внутренние. В файле схемы прописано удаление данных, после каждого запуска база чистая.
Пока отсылаю так.

shargin commented on src/main/resources/Schema.sql:
```diff
@@ -0,0 +1,96 @@
+DROP TABLE IF EXISTS "likes";
```
В проекте ожидается 2 файла schema.sql и data.sql отдельно схема и данные
По условию "В рабочем режиме H2 будет хранить данные в файле на жёстком диске. Это не позволит рабочим данным потеряться между запусками." В данном случае данные удаляются

shargin commented on src/main/resources/application.properties:
```diff
@@ -1 +1,6 @@
\ No newline at end of file
+logging.level.ru.yandex.practicum=debug
+spring.sql.init.mode=always
+spring.datasource.url=jdbc:h2:file:./db/filmorate
```
В проекте отсутствует каталог db, в таком случае можно поступить следующим образом - в каталоге db положить файл .getkeeper, т.к. каталог не пустой он попадет в репозиторий

shargin commented on src/test/resources/application.properties:
```diff
@@ -0,0 +1,6 @@
+logging.level.ru.yandex.practicum=debug
+spring.sql.init.mode=always
+spring.datasource.url=jdbc:h2:file:./src/test/db/filmorate
```
По условию "В режиме тестирования H2 будет хранить данные в памяти. **Это позволит базе быстро запуститься на время тестов и удалить все тестовые данные после их завершения", данные хранить не стоит, да еще и в каталоге с кодом

shargin commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/impl/FilmDbStorage.java:
```diff
@@ -0,0 +1,273 @@
+        }
+
+        int mpaId = rowSet.getInt("mpa_id");
+        Film film = new Film(id,
```
Тут и далее удобнее создать класс, реализующий интерфейс RowMapper, который бы преобразовывал row в объект. Его можно переиспользовать, хранить отдельно и не загромождать код.

shargin commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/impl/FilmDbStorage.java:
```diff
@@ -0,0 +1,273 @@
+                new MPA(mpaId,
+                        template.queryForObject(
+                                "SELECT \"name\" FROM \"mpa_rating\" WHERE \"mpa_id\" = " + mpaId, String.class)));
+
```
Вместо одного запроса выполняем несколько, лучше стараться писать запросы так, чтобы данные получать сразу, например, использовать join или select с несколькими таблицами

shargin commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/impl/FilmDbStorage.java:
```diff
@@ -0,0 +1,273 @@
+                    rowSet.getInt("duration"),
+                    rowSet.getInt("rate"),
+                    new MPA(mpaId,
+                            template.queryForObject(
```
И тут получим проблему N+1 на каждый объект делаем несколько запросов

shargin commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/impl/UserDbStorage.java:
```diff
@@ -0,0 +1,169 @@
+            throw new UserNotFoundException(
+                    String.format("Ошибка получения: пользователь с id=%d не найден.", id));
+        }
+        User user = new User(
```
Тут аналогично комментариям FilmDbStorage

shargin commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/impl/FilmDbStorage.java:
```diff
@@ -0,0 +1,273 @@
+@Repository("FilmDbStorage")
+@Primary
+@AllArgsConstructor
+public class FilmDbStorage implements FilmStorage {
```
Их лучше перенести в отдельный пакет, это уже dao, а не storage

shargin commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/impl/UserDbStorage.java:
```diff
@@ -0,0 +1,169 @@
+@Repository("UserDBStorage")
+@Primary
+@AllArgsConstructor
+public class UserDbStorage implements UserStorage {
```
И этот класс лучше перенести в отдельный пакет, это уже dao, а не storage. Он работает с БД

shargin commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/impl/FilmDbStorage.java:
```diff
@@ -0,0 +1,273 @@
+        }
+    }
+
+    private void insertGenres(Film film, int id) {
```
Для жанров лучше создать отдельный dao класс

IIExpanse committed:
fix: 1. Возвращен файл Data.sql. Из Schema.sql убрано удаление данных при запуске.
2. В каталог db добавлен пустой файл .gitkeep для сохранения каталога в репозитории.
3. Из тестов убрано хранение БД в файле.
4. Где возможно, десериализация объектов реализована через RowMapper.
5. Где позволяют связи между таблицами, запросы объединены через JOIN.
6. Все работающие с БД классы переименованы и вынесены в пакет dao. Из FilmDAO выделены классы для выгрузки объектов MPA и Genre.
7. Скорректирована диаграмма БД.

IIExpanse committed:
fix: 1. Возвращен файл Data.sql. Из Schema.sql убрано удаление данных при запуске.
2. В каталог db добавлен пустой файл .gitkeep для сохранения каталога в репозитории.
3. Из тестов убрано хранение БД в файле.
4. Где возможно, десериализация объектов реализована через RowMapper.
5. Где позволяют связи между таблицами, запросы объединены через JOIN.
6. Все работающие с БД классы переименованы и вынесены в пакет dao. Из FilmDAO выделены классы для выгрузки объектов MPA и Genre.
7. Скорректирована диаграмма БД.

IIExpanse committed:
fix: 1. Возвращен файл Data.sql. Из Schema.sql убрано удаление данных при запуске.
2. В каталог db добавлен пустой файл .gitkeep для сохранения каталога в репозитории.
3. Из тестов убрано хранение БД в файле.
4. Где возможно, десериализация объектов реализована через RowMapper.
5. Где позволяют связи между таблицами, запросы объединены через JOIN.
6. Все работающие с БД классы переименованы и вынесены в пакет dao. Из FilmDAO выделены классы для выгрузки объектов MPA и Genre.
7. Скорректирована диаграмма БД.

Nagash40 commented:
Андрей, высылаю новую версию с учетом комментариев.
С тестами проблема сохраняется: локально все проходит, через гит - нет.

IIExpanse committed:
fix: Названия файлов схем и данных переведены в нижний регистр.

Nagash40 commented:
После перевода названий схем и данных в нижний регистр тесты прошли. Спасибо за помощь!
